File: backend/m0_field/geometry.py
"""Field geometry: pin -> polygon, centroid, area.

The Build Brief wants a boundary drawn in under 30 seconds with no typing.
True Sentinel-2 segmentation is a scale path; for the prototype a pin becomes a
square plot of typical smallholder size, and a farmer-drawn polygon is accepted
verbatim. Both paths produce the same shape downstream.
"""

from __future__ import annotations

import math

# Mean smallholder plot in the Haryana wheat belt. Tunable per nation.
DEFAULT_PLOT_HA = 1.5

METRES_PER_DEGREE_LAT = 110_574.0


def _metres_per_degree_lng(latitude: float) -> float:
    return 111_320.0 * math.cos(math.radians(latitude))


def square_polygon_around(lat: float, lng: float, area_ha: float = DEFAULT_PLOT_HA) -> list[list[float]]:
    """Return a closed square polygon of `area_ha` centred on the pin.

    Points are [lat, lng] to match the Firestore `Field.polygon` contract.
    """
    side_m = math.sqrt(area_ha * 10_000.0)
    half_lat = (side_m / 2.0) / METRES_PER_DEGREE_LAT
    half_lng = (side_m / 2.0) / _metres_per_degree_lng(lat)

    return [
        [round(lat - half_lat, 6), round(lng - half_lng, 6)],
        [round(lat - half_lat, 6), round(lng + half_lng, 6)],
        [round(lat + half_lat, 6), round(lng + half_lng, 6)],
        [round(lat + half_lat, 6), round(lng - half_lng, 6)],
        [round(lat - half_lat, 6), round(lng - half_lng, 6)],
    ]


def centroid_of(polygon: list[list[float]]) -> dict[str, float]:
    """Arithmetic centroid of the polygon's distinct vertices."""
    points = _open_ring(polygon)
    return {
        "lat": round(sum(p[0] for p in points) / len(points), 6),
        "lng": round(sum(p[1] for p in points) / len(points), 6),
    }


def area_ha(polygon: list[list[float]]) -> float:
    """Shoelace area in hectares, projecting degrees to metres about the centroid.

    Accurate enough at field scale (sub-kilometre), where the local flat-earth
    approximation costs far less than the precision of the boundary itself.
    """
    points = _open_ring(polygon)
    if len(points) < 3:
        return 0.0

    origin_lat = sum(p[0] for p in points) / len(points)
    lat_scale = METRES_PER_DEGREE_LAT
    lng_scale = _metres_per_degree_lng(origin_lat)

    xy = [((p[1] - points[0][1]) * lng_scale, (p[0] - points[0][0]) * lat_scale) for p in points]

    total = 0.0
    for i in range(len(xy)):
        x1, y1 = xy[i]
        x2, y2 = xy[(i + 1) % len(xy)]
        total += x1 * y2 - x2 * y1

    return round(abs(total) / 2.0 / 10_000.0, 4)


def to_geojson(polygon: list[list[float]]) -> dict:
    """GeoJSON Polygon ([lng, lat] order) for Earth Engine and map layers."""
    ring = _open_ring(polygon)
    coordinates = [[p[1], p[0]] for p in ring]
    coordinates.append(coordinates[0])
    return {"type": "Polygon", "coordinates": [coordinates]}


def _open_ring(polygon: list[list[float]]) -> list[list[float]]:
    """Drop the duplicated closing vertex if present."""
    if len(polygon) > 1 and polygon[0] == polygon[-1]:
        return polygon[:-1]
    return polygon


def is_simple(polygon: list[list[float]]) -> bool:
    """True when no two edges of the boundary cross.

    A farmer dragging a finger across their own line makes a bow-tie, and the
    shoelace formula quietly sums the two lobes with opposite winding: a crossed
    square does not merely under-report, it returns 0.0 ha. Every per-hectare
    number downstream — seed, fertiliser, yield, credit — divides by that.
    Better to refuse the boundary than to advise on it.
    """
    ring = _open_ring(polygon)
    count = len(ring)
    if count < 4:  # a triangle cannot cross itself
        return True

    for i in range(count):
        a1, a2 = ring[i], ring[(i + 1) % count]
        # Start at i+2: neighbouring edges legitimately share a vertex.
        for j in range(i + 2, count):
            # The last edge wraps to vertex 0, so it neighbours edge 0.
            if i == 0 and j == count - 1:
                continue
            if _segments_cross(a1, a2, ring[j], ring[(j + 1) % count]):
                return False
    return True


def _segments_cross(p1, p2, q1, q2) -> bool:
    """Proper intersection test via the four orientation signs."""
    d1 = _orientation(q1, q2, p1)
    d2 = _orientation(q1, q2, p2)
    d3 = _orientation(p1, p2, q1)
    d4 = _orientation(p1, p2, q2)

    if ((d1 > 0) != (d2 > 0)) and ((d3 > 0) != (d4 > 0)):
        return True

    # Collinear overlap: a segment doubling back along its neighbour.
    return (
        (d1 == 0 and _on_segment(q1, q2, p1))
        or (d2 == 0 and _on_segment(q1, q2, p2))
        or (d3 == 0 and _on_segment(p1, p2, q1))
        or (d4 == 0 and _on_segment(p1, p2, q2))
    )


def _orientation(a, b, c) -> float:
    """Cross product sign: >0 left turn, <0 right turn, 0 collinear."""
    value = (b[1] - a[1]) * (c[0] - b[0]) - (b[0] - a[0]) * (c[1] - b[1])
    if abs(value) < 1e-12:
        return 0.0
    return value


def _on_segment(a, b, point) -> bool:
    return (
        min(a[0], b[0]) <= point[0] <= max(a[0], b[0])
        and min(a[1], b[1]) <= point[1] <= max(a[1], b[1])
    )

File: app/lib/features/field/domain/geometry.dart
/// Client-side mirror of `backend/m0_field/geometry.py`.
///
/// The Build Brief wants the boundary outlined in under 30 seconds, so the app
/// draws the polygon the instant the pin moves rather than waiting on a round
/// trip. The backend still recomputes it authoritatively; this only has to
/// agree well enough to be the same square.
library;

import 'dart:math' as math;

const double defaultPlotHa = 1.5;
const double _metresPerDegreeLat = 110574.0;

double _metresPerDegreeLng(double latitude) =>
    111320.0 * math.cos(latitude * math.pi / 180.0);

/// A closed square polygon of [areaHa] centred on the pin, as [lat, lng] pairs.
List<List<double>> squarePolygonAround(
  double lat,
  double lng, {
  double areaHa = defaultPlotHa,
}) {
  final sideM = math.sqrt(areaHa * 10000.0);
  final halfLat = (sideM / 2.0) / _metresPerDegreeLat;
  final halfLng = (sideM / 2.0) / _metresPerDegreeLng(lat);

  double r(double v) => double.parse(v.toStringAsFixed(6));

  return [
    [r(lat - halfLat), r(lng - halfLng)],
    [r(lat - halfLat), r(lng + halfLng)],
    [r(lat + halfLat), r(lng + halfLng)],
    [r(lat + halfLat), r(lng - halfLng)],
    [r(lat - halfLat), r(lng - halfLng)],
  ];
}

/// Arithmetic centroid of the polygon's distinct vertices.
({double lat, double lng}) centroidOf(List<List<double>> polygon) {
  final points = openRing(polygon);
  if (points.isEmpty) return (lat: 0, lng: 0);
  final lat = points.map((p) => p[0]).reduce((a, b) => a + b) / points.length;
  final lng = points.map((p) => p[1]).reduce((a, b) => a + b) / points.length;
  return (lat: lat, lng: lng);
}

/// Shoelace area in hectares, projecting degrees to metres about the centroid.
double areaHaOf(List<List<double>> polygon) {
  final points = openRing(polygon);
  if (points.length < 3) return 0;

  final originLat =
      points.map((p) => p[0]).reduce((a, b) => a + b) / points.length;
  final lngScale = _metresPerDegreeLng(originLat);

  final xy = points
      .map((p) => (
            x: (p[1] - points[0][1]) * lngScale,
            y: (p[0] - points[0][0]) * _metresPerDegreeLat,
          ))
      .toList();

  var total = 0.0;
  for (var i = 0; i < xy.length; i++) {
    final a = xy[i];
    final b = xy[(i + 1) % xy.length];
    total += a.x * b.y - b.x * a.y;
  }

  return (total.abs() / 2.0) / 10000.0;
}

/// Drop the duplicated closing vertex if present.
List<List<double>> openRing(List<List<double>> polygon) =>
    polygon.length > 1 &&
            polygon.first[0] == polygon.last[0] &&
            polygon.first[1] == polygon.last[1]
        ? polygon.sublist(0, polygon.length - 1)
        : polygon;

/// True when no two edges of the boundary cross.
///
/// Mirrors `is_simple` in backend/m0_field/geometry.py. A farmer dragging
/// across their own line makes a bow-tie, and the shoelace formula sums the two
/// lobes with opposite winding: a crossed square reports 0.0 ha, not a wrong
/// one. Catching it here saves the round-trip and explains the problem while
/// the boundary is still on screen.
bool isSimple(List<List<double>> polygon) {
  final ring = openRing(polygon);
  final count = ring.length;
  if (count < 4) return true; // a triangle cannot cross itself

  for (var i = 0; i < count; i++) {
    final a1 = ring[i];
    final a2 = ring[(i + 1) % count];
    // Start at i+2: neighbouring edges legitimately share a vertex.
    for (var j = i + 2; j < count; j++) {
      // The last edge wraps to vertex 0, so it neighbours edge 0.
      if (i == 0 && j == count - 1) continue;
      if (_segmentsCross(a1, a2, ring[j], ring[(j + 1) % count])) return false;
    }
  }
  return true;
}

bool _segmentsCross(
  List<double> p1,
  List<double> p2,
  List<double> q1,
  List<double> q2,
) {
  final d1 = _orientation(q1, q2, p1);
  final d2 = _orientation(q1, q2, p2);
  final d3 = _orientation(p1, p2, q1);
  final d4 = _orientation(p1, p2, q2);

  if ((d1 > 0) != (d2 > 0) && (d3 > 0) != (d4 > 0)) return true;

  return (d1 == 0 && _onSegment(q1, q2, p1)) ||
      (d2 == 0 && _onSegment(q1, q2, p2)) ||
      (d3 == 0 && _onSegment(p1, p2, q1)) ||
      (d4 == 0 && _onSegment(p1, p2, q2));
}

/// Cross product sign: >0 left turn, <0 right turn, 0 collinear.
double _orientation(List<double> a, List<double> b, List<double> c) {
  final value = (b[1] - a[1]) * (c[0] - b[0]) - (b[0] - a[0]) * (c[1] - b[1]);
  return value.abs() < 1e-12 ? 0.0 : value;
}

bool _onSegment(List<double> a, List<double> b, List<double> point) =>
    point[0] >= (a[0] < b[0] ? a[0] : b[0]) &&
    point[0] <= (a[0] > b[0] ? a[0] : b[0]) &&
    point[1] >= (a[1] < b[1] ? a[1] : b[1]) &&
    point[1] <= (a[1] > b[1] ? a[1] : b[1]);
